### PR TITLE
Add support for converting with-items

### PR DESCRIPTION
**Pull request by blag**
This PR adds support for converting Mistral's `with-items` and `concurrency` task attributes.

See [this page](https://docs.openstack.org/mistral/latest/user/wf_lang_v2.html#processing-collections) of the Mistral docs for some examples.

```yaml
with-items:
  - i in <% $.yaql_data %>

with-items:
  - i in <% $.yaql_data %>
  - j in <% $.more_data %>
  - k in <% $.all_the_data %>

with-items: i in <% $.yaql_data %>
concurrency: 2

with-items: i in {{ _.jinja_data }}

with-items: i in [0, 1, 2, 3]
```

should produce the following in orquesta

```yaml
with:
  items: i in <% ctx().yaql_data %>

with:
  items: i, j, k in <% zip($.yaql_data, $.more_data, $.all_the_data) %>

with:
  items: i in <% $.yaql_data %>
  concurrency: 2

with:
  items: i in {{ ctx().jinja_data }}

with:
  items: i in <% [0, 1, 2, 3] %>
```

There is quite a bit of code churn in the pack tests, and it was a little difficult for me to extract the churn from the additional tests, so I recommend looking at the end results than the commit diff for that. I added quite a few additional fixtures for the integration tests for the pack client.

Interpreting and converting the `with-items` list correctly requires that all items in the list utilize the same expression format (eg: either all static/Jinja expressions or all static/YAQL expressions), however, it does not validate that.

Also, this _does_ support syntax like:

```yaml
with-items: i in [0, 1, 2, 3]
```

and converts it to:

```yaml
with:
  items: i in <% [0, 1, 2, 3] %>
```

because we have some workflows that utilize that. I'm not 100% sure it's valid Mistral syntax, but since Mistral does seem to run it just fine I wanted to convert it to a more proper format that Orquesta can run just fine as well.

### CODE REVIEW

File: orquestaconvert/workflows/base.py [hunk 1]
```diff
@@ -251,11 +256,102 @@ def convert_action(self, m_action):
...
+        o_action = MISTRAL_ACTION_CONVERSION_TABLE.get(m_action, m_action)
+
+        kwargs = {'item_vars': self.task_with_item_vars}
+        return MixedExpressionConverter.convert_string(o_action, **kwargs)
+
+    def convert_input(self, input_):
+        kwargs = {'item_vars': self.task_with_item_vars}
+        return MixedExpressionConverter.convert_dict(input_, **kwargs)
+
+    def convert_with_items_expr(self, expression, expr_converter):
```
Comment (nmaludy) [inline orquestaconvert/workflows/base.py]: this is a gigantic function and should probably be split up for easier testing and readability

File: orquestaconvert/workflows/base.py [hunk 2]
```diff
@@ -269,14 +365,28 @@ def convert_tasks(self, mistral_wf_tasks, expr_converter, force=False):
                                                    " that is not supported in orquesta.").
                                                   format(task_name, attr))
 
+            if m_task_spec.get('with-items'):
```
Comment (nmaludy) [inline orquestaconvert/workflows/base.py]: should probably be its own function

File: orquestaconvert/expressions/mixed.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+from orquestaconvert.expressions import ExpressionConverter
+from orquestaconvert.expressions.base import BaseExpressionConverter
+from orquestaconvert.utils import type_utils
+
+
+JINJA_EXPR_RGX = re.compile(r'(?P<expr>(?:{{)\s*.+?\s*(?:}}))')
+YAQL_EXPR_RGX = re.compile(r'(?P<expr>(?:<%)\s*.+?\s*(?:%>))')
+
+
+class MixedExpressionConverter(BaseExpressionConverter):
```
Comment (nmaludy) [inline orquestaconvert/expressions/mixed.py]: what is a "mixed" expression?

File: orquestaconvert/expressions/mixed.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import warnings
+
+import ruamel.yaml.comments
+
+from orquestaconvert.expressions import ExpressionConverter
+from orquestaconvert.expressions.base import BaseExpressionConverter
+from orquestaconvert.utils import type_utils
+
+
+JINJA_EXPR_RGX = re.compile(r'(?P<expr>(?:{{)\s*.+?\s*(?:}}))')
```
Comment (nmaludy) [inline orquestaconvert/expressions/mixed.py]: explanations of what these regexes are and what they're matching would be nice

File: tests/base_test_case.py [hunk 5]
```diff
@@ -64,3 +66,71 @@ def setUp(self):
...
+            'mistral-with-items-static-list.yaml',
+            'mistral-with-items-mixed-list.yaml',
+            'mistral-with-items-yaql.yaml',
+            'mistral-with-items-yaml-list.yaml',
+            'mistral-with-items-yaql-list.yaml',
+            'mistral-with-items-concurrency.yaml',
+        ]
+        self.action_files = self.action_failing_files + self.action_passing_files
+
+        self.pristine_action_wfs = {
```
Comment (nmaludy) [inline tests/base_test_case.py]: these look like the same thing over and over, should probably be a function

File: tests/base_test_case.py [hunk 6]
```diff
@@ -64,3 +66,71 @@ def setUp(self):
...
+        super(BasePackClientRunTestCase, self).setUp()
+
+        # The pristine actions directory - this is not touched
+        self.p_actions_dir = os.path.join('tests', 'fixtures', 'pack', 'pristine_actions')
+        # The Orquesta actions directory - this is not touched
+        self.o_actions_dir = os.path.join('tests', 'fixtures', 'pack', 'o_actions')
+        # The Mistral actions directory - this is a copy of pristine actions
+        self.m_actions_dir = os.path.join('tests', 'fixtures', 'pack', 'actions')
+
+        self.action_failing_files = [
```
Comment (nmaludy) [inline tests/base_test_case.py]: thinking these should be constants above the class?

File: orquestaconvert/expressions/mixed.py (hunk 3 above)
Comment (blag) [reply orquestaconvert/expressions/mixed.py]: Explained. 👍

File: orquestaconvert/workflows/base.py (hunk 2 above)
Comment (blag) [reply orquestaconvert/workflows/base.py]: Yep. Moved.

**Pull request merged**